Unify message checks for GPU process

rniwa opened the pull request:
#### eaf544ce68542ca6c1bc390a04ecc8cec39b3c5f
<pre>
Unify message checks for GPU process
<a href="https://bugs.webkit.org/show_bug.cgi?id=278130">https://bugs.webkit.org/show_bug.cgi?id=278130</a>

Reviewed by Sihui Liu and Geoffrey Garen.

Unify the implementation of MESSAGE_CHECK in GPU process and void duplicating the logic to kill Web process
in each macro since GPUConnectionToWebProcess::markCurrentlyDispatchedMessageAsInvalid already does that.

* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::getPixelBuffer):
(WebKit::RemoteImageBuffer::getPixelBufferWithNewMemory):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::releaseImageBuffer):
(WebKit::RemoteRenderingBackend::releaseRemoteImageBufferSet):
(WebKit::RemoteRenderingBackend::cacheFont):
(WebKit::RemoteRenderingBackend::cacheFontCustomPlatformData):
(WebKit::RemoteRenderingBackend::releaseRenderingResource):
(WebKit::RemoteRenderingBackend::prepareImageBufferSetsForDisplay):
(WebKit::RemoteRenderingBackend::prepareImageBufferSetsForDisplaySync):
(WebKit::RemoteRenderingBackend::markSurfacesVolatile):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandEncoder.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundleEncoder.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteTexture.cpp:
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.cpp:
(WebKit::RemoteMediaRecorder::audioSamplesStorageChanged):
(WebKit::RemoteMediaRecorder::audioSamplesAvailable):
* Source/WebKit/Platform/IPC/Connection.h:

Canonical link: <a href="https://commits.webkit.org/282321@main">https://commits.webkit.org/282321@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e55837e3ec57c3cba4896484ecbf85d69f37cc4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62808 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42164 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15404 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66828 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13412 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49851 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13696 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/50661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9260 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65877 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39202 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/54408 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/31340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/35906 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12288 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/57452 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/12071 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68523 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6754 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/11715 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/57980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6786 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54463 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5645 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9462 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37984 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/39064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/40175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/38806 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandEncoder.cpp:
```diff
@@ -38,17 +39,14 @@
-        if (auto connection = m_gpuConnectionToWebProcess.get()) \
-            connection->terminateWebProcess(); \
-        return; \
+    if (auto didPassMessageCheck = !!(assertion)) { \
```
Maybe the !! is confusing me, but doesn't this do nothing if the assertion is false?

mwyrzykowski replied:
`!!` being effectively a conversion to bool operator, it would appear to be the case indeed which seems unintentional

rniwa replied:
Oh oops, you're right. I meant to say !(assertion) here.

geoffreygaren commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandEncoder.cpp:
```diff
@@ -38,17 +39,14 @@
-        return; \
+    if (auto didPassMessageCheck = !!(assertion)) { \
+        auto connection = m_gpuConnectionToWebProcess.get(); \
+        if (!connection) \
```
And doesn't this do nothing if the connection is null (when we would prefer to kill the WebContent process instead)?

geoffreygaren replied:
I guess we can't kill the WebContent process if we have no connection. Maybe we should kill the GPU process in that case, though? We're definitely Having a Bad Time (TM) if we receive an invalid message from a process we can't identify. If our goal is to fail closed, I think we do need to kill the GPU process.

rniwa replied:
But there could be a race, right? WebContent can send an IPC to GPU process shortly before GPU process had a chance to process the last IPC.

mwyrzykowski replied:
Indeed and bringing down the entire GPU process would be quite problematic in that case

geoffreygaren replied:
I guess if m_gpuConnectionToWebProcess only becomes null when the WebContent process exits, maybe that's just as good as killing the WebContent process, so there's no need to do anything?

szewai commented on Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.cpp:
```diff
@@ -37,7 +37,14 @@
+        auto connection = m_gpuConnectionToWebProcess.get(); \
+        if (!connection) \
+            return; \
+        MESSAGE_CHECK_BASE(!didFailAssertion, &connection->connection()); \
```
Maybe we can add null check for connection in `MESSAGE_CHECK_BASE` instead? It seems more robust than adding the null check for every `MESSAGE_CHECK`

rniwa replied:
I don't think we can do that because MESSAGE_CHECK_BASE takes Connection as an argument, not GPUConnectionToWebProcess.

szewai replied:
I mean something like:
```
#define MESSAGE_CHECK(assertion) MESSAGE_CHECK_BASE(assertion, connection())
IPC::Connection* RemoteMediaRecorder::connection() {
    return m_gpuConnectionToWebProcess ? &m_gpuConnectionToWebProcess->connection() : nullptr;
}
void RemoteMediaRecorder::audioSamplesAvailable(MediaTime time, uint64_t numberOfFrames)
{
    MESSAGE_CHECK(m_ringBuffer);
    ....
}
...
#define MESSAGE_CHECK_COMPLETION_BASE(assertion, connection, completion) do { \
    if (UNLIKELY(!(assertion))) { \
        RELEASE_LOG_FAULT(IPC, __FILE__ " " CONNECTION_STRINGIFY_MACRO(__LINE__) ": Invalid message dispatched %s", WTF_PRETTY_FUNCTION); \
        if (connection)
            connection->markCurrentlyDispatchedMessageAsInvalid(); \
        { completion; } \
        return; \
    } \
} while (0)
```

rniwa replied:
No, that would be racy since `m_gpuConnectionToWebProcess` is a `ThreadSafeWeakPtr`.

rniwa replied:
I'd remove the `connection` local variable from `RemoteMediaRecorder::audioSamplesStorageChanged` as we discussed in person.

geoffreygaren reviewed (approved):
r=me

rniwa committed:
Unify message checks for GPU process
https://bugs.webkit.org/show_bug.cgi?id=278130

Reviewed by Sihui Liu and Geoffrey Garen.

Unify the implementation of MESSAGE_CHECK in GPU process and void duplicating the logic to kill Web process
in each macro since GPUConnectionToWebProcess::markCurrentlyDispatchedMessageAsInvalid already does that.

* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::getPixelBuffer):
(WebKit::RemoteImageBuffer::getPixelBufferWithNewMemory):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::releaseImageBuffer):
(WebKit::RemoteRenderingBackend::releaseRemoteImageBufferSet):
(WebKit::RemoteRenderingBackend::cacheFont):
(WebKit::RemoteRenderingBackend::cacheFontCustomPlatformData):
(WebKit::RemoteRenderingBackend::releaseRenderingResource):
(WebKit::RemoteRenderingBackend::prepareImageBufferSetsForDisplay):
(WebKit::RemoteRenderingBackend::prepareImageBufferSetsForDisplaySync):
(WebKit::RemoteRenderingBackend::markSurfacesVolatile):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandEncoder.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundleEncoder.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteTexture.cpp:
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.cpp:
(WebKit::RemoteMediaRecorder::audioSamplesStorageChanged):
(WebKit::RemoteMediaRecorder::audioSamplesAvailable):
* Source/WebKit/Platform/IPC/Connection.h:

Canonical link: https://commits.webkit.org/282321@main
